updates

moham96 opened the pull request:
- parse settings in pushbutton.py
- trigger pushbutton.py using the keyboard in platforms without gpios 
- control tasmota with multiple relays 
- make pygame plays nicely with devices running pulseaudio 
- refactoring and cleanups

shivasiddharth commented on Requirements/GassistPi-system-requirements.txt:
```diff
@@ -12,5 +12,4 @@ python3-venv
 python3
 python3-pip
 socat
-mpg123
```
Needed for "say" function (Text to speech converter).

moham96 replied:
not need, I converted the say function to use pygame instead [L185!](https://github.com/moham96/GassistPi/blob/0212d52a9ab0282d746753e2e4f8972fe58b9816/src/actions.py#L185)

moham96 commented:
@shivasiddharth, well the trigger is not the enter key, I've set it to F12 for my setup (run the service again and press f12), but the user can set it to anything he/she wants:
-Run the getkeystroke.py script inside the src folder `python GassistPi/src/getkeystroke.py`
-It will tell you to press a key, so press the key you want to be the trigger, you'll get the ASCII value of the key.
-Next, open the pushbutton.py and change the triggerkey variable to the ascii value you got [L116!](triggerkey=201)
-Now restart the service and your trigger key should work wherever your cursor is
(This process might seem a little bit convoluted,so probably my next update will be about combining the user settings in a single file and let the assistant set them "Hey Google, set my trigger key to F9")

PS: If you decided to merge my commit's there will be no need for the patch-to-parse-project-id-and-model-id-from-file and osmc branches so you can delete them

shivasiddharth commented:
Fine thanks.

shivasiddharth commented:
I did not notice that it was pull request to the master. Will test it over the weekend with other boards and then merge it with the master.

moham96 commented:
@shivasiddharth  tell me if anything comes up in your tests.

shivasiddharth commented:
Based on trying the installation in armbian on Orangepi:
1. In https://github.com/moham96/GassistPi/blob/master/scripts/gassist-installer.sh#L53 $projiid needs to be changed to $project_id
2. pygame installation terminates with `Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-0kqb32zp/pygame/` error

shivasiddharth commented:
Also, i guess we need to move the RPi.GPIO from the pip requirements file to the body of main installer. First we need to check if the board model is Raspberry Pi and then install the RPi.GPIO else skip the package installation.

moham96 commented:
@shivasiddharth  How about using [libvlc](https://wiki.videolan.org/python_bindings) This way we can ditch both Pygame and Mpv in the entire code, it's pure python and we have audio controls
Can you test the following code in orangepi?
```
import vlc
#url = 'http://www.radiofeeds.co.uk/bbcradio2.pls'
url="/home/mohammad/GassistPi/sample-audio-files/Fb.wav"
Instance = vlc.Instance("--verbose 2".split())
player = Instance.media_player_new()
Media_list = Instance.media_list_new([url])
list_player = Instance.media_list_player_new()
list_player.set_media_player(player)
list_player.set_media_list(Media_list)
list_player.play()
while list_player.get_state() != vlc.State.Stopped:
    pass
```
Install python-vlc using pip and vlc using apt, then run the code above
try changing url to different files (mp3, wav, radio pls or m3u)
see If there is any problem is playback
also see if this solves #187 
you can control audio using 
```
player.audio_get_volume()
player.audio_set_volume(70)
player.audio_set_mute(1)
```

shivasiddharth commented:
Two reasons why i moved to mpv:
1. Alexa-avs-sample-app uses vlc.
2. gmusicapi streams could not be played in vlc.

If someone is running alexa alongside this project, use the command to stop media, it would close vlc subprocess and hence crash alexa as well. Last time when i tried, gmusicapi streams could not be played using vlc.

Can you please test the gmusicapi streams with vlc, while i shall try and port the Alexa to Alexa SDK.

moham96 commented:
There is no vlc subprocess to stop, python-vlc is pure python wrap around the C libvlc, so there will not be a vlc process running

shivasiddharth commented:
Let me test this with gmusicapi and get back then

moham96 commented:
@shivasiddharth I'm don't fully understand. I've already updated main.py and pushbutton.py to parse the info from the info file and didn't have an issue. What exactly is the issue you faced?

shivasiddharth commented:
I had earlier made a post on this in the SDK G+ forums. https://plus.google.com/108883079869581634525/posts/8pePUoyeECg when you manually run the main.py, if you include the project_id, then it creates a new device instance everytime you run it manually. Say for example someone is testing something by manually running the main.py, if it crashes and the person again starts the main.py then it will create a second instance, so on and so forth. That is why, i removed the project_id args in the service file and the readme doc.

moham96 commented:
@shivasiddharth I just tasted my main.py and It's not registering new instances, I deleted all my instances in the app and ran the script again so the script registered a new instance and on repeating it keeps using old instance. Does the problem happen always or randomly ? can you start fresh with my latest commits in the master branch(my fork).Be sure to delete the info file (~/gassistant-credentials.info.
Maybe this problem is fixed in the versions of Google packages that I'm using. I think I updated them some time ago so they're probably not the same as you specified in the installer.
I will include the versions of packages I'm using so you could install the same versions and see if the problem goes away.
This is the output of `pip freeze`:

```
aftership==0.2
appdirs==1.4.3
beautifulsoup4==4.6.0
cachetools==2.0.1
certifi==2018.1.18
cffi==1.11.5
chardet==3.0.4
click==6.7
decorator==4.2.1
feedparser==5.2.1
future==0.16.0
futures==3.1.1
gmusicapi==11.0.0
google-api-python-client==1.6.5
google-assistant-grpc==0.1.0
google-assistant-library==0.1.1
google-assistant-sdk==0.4.2
google-auth==1.3.0
google-auth-httplib2==0.0.3
google-auth-oauthlib==0.2.0
googleapis-common-protos==1.5.3
googletrans==2.2.0
gpsoauth==0.4.1
grpcio==1.10.0
gTTS==1.2.2
gTTS-token==1.1.1
httplib2==0.10.3
idna==2.6
jedi==0.11.1
keyboard==0.11.0
kodi-json==1.0.0
lxml==4.1.1
MechanicalSoup==0.10.0
mock==2.0.0
monotonic==1.4
mutagen==1.40.0
netifaces==0.10.6
oauth2client==4.1.2
oauthlib==2.0.6
pafy==0.5.4
parso==0.1.1
pathlib2==2.3.0
pbr==3.1.1
proboscis==1.2.6.0
protobuf==3.5.1
psutil==5.4.3
pushbullet.py==0.11.0
pyaml==17.12.1
pyasn1==0.4.2
pyasn1-modules==0.2.1
PyAudio==0.2.11
PyChromecast==2.0.0
pycparser==2.18
pycryptodomex==3.4.12
pygame==1.9.3
python-dateutil==2.6.1
python-magic==0.4.15
python-vlc==3.0.102
PyYAML==3.12
requests==2.18.4
requests-oauthlib==0.8.0
RPi.GPIO==0.6.3
rsa==3.4.2
six==1.11.0
sounddevice==0.3.10
tenacity==4.8.0
thonny==2.1.16
uritemplate==3.0.0
urllib3==1.22
validictory==1.1.2
websocket-client==0.47.0
xlib==0.21
youtube-dl==2018.2.26
zeroconf==0.20.0
```

shivasiddharth commented:
You are on 0.4.2 SDK. The issue started with 0.4.3.

moham96 commented:
@shivasiddharth Ok, I tried 0.4.3 and 0.4.4 and still I'm not getting multiple devices in my app, And I can unlink from the app!

shivasiddharth commented:
The issue seems to be a backend issue which Google have fixed now. Fair enough !

moham96 commented:
@shivasiddharth  can you give a try to [https://github.com/moham96/GassistPi/tree/dev](https://github.com/moham96/GassistPi/tree/dev) I only updated main.py.
Sometimes I get 403 forbidden error when playing google music even with mpv. Do you have the same behavior?
also there are some bits of the code that are ambiguous to me:
- What is the practical use of `trackchange.py` file? I'm sure we can find a way to ditch it's usage.
- Why do you use json files like `~/songs.json`? this way if I added songs to my google account I cannot play them, I need to manually delete this file to reload the songs list

shivasiddharth commented:
I shall check that out. i have a  lua script for mpv that will detect the end of playback of the file and upon reaching the end, it will execute the python script (trackchange.py). This is universal for all media playback (contents of the trackchange.py will be overwritten in the main.py based upon the requested media). Otherwise, you need to write separate EOF detection events for different types of media and create a threaded process. If you dont have the local database then everytime pi needs to fetch the list from network which will cause a delay. You actually dont need to manually delete the local database, i have already placed a refresh command in the main.py (https://github.com/shivasiddharth/GassistPi/blob/master/src/main.py#L465) just by giving the refresh command the local database will be updated. I forgot to mention this in readme.

moham96 commented:
@shivasiddharth aha, so you write the track index that is playing inside `~/.gmusicsongsplayer.json` for example and when mpv finishes playing it executes `end.lua` which in turn execute `trackchange.py` which will run the function again so it will play the next track and increment the id(correct me if I'm wrong). 
So we need to accomplish the same thing inisde libvlc, I think libvlc already has an event manager that we might be able to use, or since libvlc has a playlist player we can load the entire playlist to libvlc and so libvlc will do all the work

shivasiddharth commented:
Yup! Thats what i have done. 
You can refer to this for the vlc event manager https://github.com/Tom-Archer/gmusicaiy/blob/master/playscroll.py In that sample, the player and event manager instance has been embedded with the gmusic codes. Ideally, vlc player and event instance can be declared under a separate function and the same can be called for  different media playbacks. That would require minimal edits to the existing functions.

shivasiddharth commented:
I am yet to try your branch, was actually going through your files. Looks good but do we really need to make things complicated by having multiple files/scripts? So far, people have found this project easy to use because of single point of edits. By using the config.yaml and utils.py we are introducing more room for user input error.
So whats the proposed advantage of using the config.yaml ? Moreover it will make all the existing guides void due to the change in file structure. 
Let me know your thoughts.

moham96 commented:
Also one of the main reasons that I created the config.yaml file is so that I can add it to the .gitignore file.I made two commits in the past few days that contained my google account passwords inside the actions.py, I added my credentials and forgot to remove them before committing. having a separate config file would eliminate such mistakes.

shivasiddharth commented:
Even i have done the same thing a couple of times lol.. This is a community project, so it has nothing to do with my vision. Once you are done with your edits, generate a pull req to the dev branch will test it and merge. we can move this conv over to the gitter. If you need a hand in anything feel free to drop in a msg there.

shivasiddharth commented:
You can also add the scripts that i have used in the other project (https://github.com/shivasiddharth/Assistants-Pi) to allow users to the choose the audio setup and test their audio.

aandroide commented:
if you want I can give you a hand starting to add this important update that you have discovered of libvlc In which solves audio problems with the aiy kit.

shivasiddharth commented:
@aandroide , i am changing the codes for google music and youtube. i will release it with vlc

moham96 commented:
@aandroide, unfortunately, I currently have no time to work on this, hopefully, in the next week I will be able to finish the work.
I solved a lot of problems and refactored a lot of code, just need to do some tests and merge.

aandroide commented:
ok guys, good job

moham96 commented:
@aandroide well it seems that @shivasiddharth already made the changes so you should be able to use them if you update.

aandroide commented:
I have updated and there are several errors that I am correcting shortly I make the pull request

aandroide commented:
instead for main.py I noticed that you have to give permissions to the 2 json files of the mediaplayer otherwise it does not work

shivasiddharth commented:
@aandroide , i have fixed the errors in pushbutton. Can you give it a try

aandroide commented:
always in pushbutton.py, missing the piece of code for the stream of music from youtube that in my opinion, it should be:
`if 'stream'.lower () in str (usrcmd) .lower ():
                        vlcplayer.stop_vlc ()
                        if os.path.isfile ("/ home / pi / .trackqueue.json"):
                            os.system ('rm /home/pi/.trackqueue.json')
                            os.system ('echo "from actions import youtubeplayer \ n \ n" >> /home/pi/.trackqueue.json')
                            os.system ('echo "youtubeplayer () \ n" >> /home/pi/.trackqueue.json')
                            if 'autoplay'.lower () in str (usrcmd) .lower ():
                                YouTube_Autoplay (str (usrcmd) flower setting ())
                            else:
                                YouTube_No_Autoplay (str (usrcmd) flower setting ())
                        else:
                            os.system ('echo "from actions import youtubeplayer \ n \ n" >> /home/pi/.trackqueue.json')
                            os.system ('echo "youtubeplayer () \ n" >> /home/pi/.trackqueue.json')
                            if 'autoplay'.lower () in str (usrcmd) .lower ():
                                YouTube_Autoplay (str (usrcmd) flower setting ())
                            else:
                                YouTube_No_Autoplay (str (usrcmd) flower setting ())
                        return continue_conversation`

aandroide commented:
also the problem remained with:
#if you uncomment does not print anything and gets stuck after some requested command
# logging.basicConfig (filename = '/ tmp / GassistPi.log', level = logging.DEBUG,
# format = '% (asctime) s% (levelname) s% (name) s% (message) s')
# Logger = logging.getLogger (__ name__)
crashes the program is still I did not understand the reason.
Just ask 2 or 3 times tell me a joke and go crashing, if you comment on those lines, it works very well and also prints the logs when you activate or close the microphone.

shivasiddharth commented:
The custom commands in pushbutton were not working because of the location of return continue_conversation. It was in a wrong position in all of the commands. I have fixed all the pushbutton related errors. I am unable to replicate the error due to the log file.

aandroide commented:
if you run pushbutton.py manually, you'll realize that if you enable the log it will never print you this:
![cattura](https://user-images.githubusercontent.com/14141574/40890655-efc16b40-6779-11e8-9621-8a6c3e72ccda.PNG)

shivasiddharth commented:
Yes. So by commenting those lines, you can see these? But do we still get the log file ?

shivasiddharth commented:
Ok, i have commented those lines. So now all the issues with pushbutton should be resolved.

aandroide commented:
the log file is obvious that we do not get it if we comment on those lines, but besides the fact that you do not see those lines, after 2 or 3 times you ask the assistant to tell you a joke, it goes crashing.
While it does not do so if we disable the log.
I repeat that I still can not understand why.
PS: I checked your corrections and it does not work. have you read my suggestions in previous posts regarding the stream that is missing in pushbutton?

aandroide commented:
@shivasiddharth  but did you really check all the code?
apparte that crashes immediately but you also told a user that everything is ok.
I saw that you commented on the lines that I told you but also in the end you have to replace:
if __name__ == '__main__':
    try:
        main ()
    except Exception as error:
        logger.exception (error)

with:
if __name__ == '__main__':
    main ()

shivasiddharth commented:
Yes i checked everything seems to be working fine. I checked the stream command, it works. Sometimes, if you have used up the API quota, stream will crash without working.

aandroide commented:
@shivasiddharth 
I'm sorry, but how does the stream work if the pushbutton.py code is missing the declaration in the stream?
![cattura](https://user-images.githubusercontent.com/14141574/40891232-28b37b02-6782-11e8-8617-ab43a22af4c5.PNG)

aandroide commented:
last problem encountered:
the stop button does not work either in main.py or pushbutton.py

aandroide commented:
@moham96  hi, can you stop the music being played by pressing the physical button?
